snooty/postprocess.py

```diff
@@ -532,11 +533,31 @@ class TabsSelectorHandler(Handler):
     def __init__(self, context: Context) -> None:
         super().__init__(context)
         self.selectors: Dict[str, List[Dict[str, MutableSequence[n.Text]]]] = {}
+        self.stack: List[str] = []
+        self.scanned_pattern: List[str] = []
+        self.target_pattern = ["tabs", "tabs", "procedure"]
+
+    def scan_for_pattern(self, fileid_stack: FileIdStack, node: n.Node) -> None:
+        starting_point = 0
+        target_pattern_len = len(self.target_pattern)
+        if len(self.scanned_pattern) > 0:
+            reverse_list = list(reversed(self.scanned_pattern))
+            for item in reverse_list:
+                if item == self.target_pattern[starting_point]:
+                    starting_point += 1
+                if starting_point >= target_pattern_len:
+                    self.context.diagnostics[fileid_stack.current].append(
+                        InvalidNestedTabStructure(" ".join(reverse_list), node.start[0])
+                    )
+                    return
 
     def enter_node(self, fileid_stack: FileIdStack, node: n.Node) -> None:
         if not isinstance(node, n.Directive):
             return
 
+        self.scanned_pattern = []
+        self.stack.append(node.name)
```

Comment:
By resetting the pattern when we enter a new node, and scanning (line 598) when exiting after popping the stack, there's a chance for the pattern to be messed up if the `procedure` is in the same level as a different directive. For example:

```
.. tabs-platforms::

   .. tab::
      :tabid: windows

      .. tabs::

         .. tab::
            :tabid: homebrew

            .. procedure::

               .. step::

                  foo

            .. note::

               Wow
```

In the example, `note` would reset the pattern before it has a chance to exit the node for the tabs, preventing the pattern to be flagged for `procedure`.

We might want to scan for the pattern sometime between adding the node to `self.scanned_pattern` and popping it out.